sockeye/inference.py

```diff
@@ -1326,13 +1362,32 @@ def _beam_search(self,
 
             # (7) determine which hypotheses in the beam are now finished
             finished = ((best_word_indices == C.PAD_ID) + (best_word_indices == self.vocab_target[C.EOS_SYMBOL]))
-            if mx.nd.sum(finished).asscalar() == self.batch_size * self.beam_size:  # all finished
-                break
+
+#            self._print_beam(sequences, scores_accumulated, finished, inactive, t)
+
+            if self.beam_search_stop == 'first' and self.batch_size == 1:
+                # TODO: doesn't work for batching
+                if mx.nd.sum(finished).asscalar() > 0:
+                    break
+            else:
+                if mx.nd.sum(finished).asscalar() == self.batch_size * self.beam_size:  # all finished
+                    break
 
             # (8) update models' state with winning hypotheses (ascending)
             for ms in model_states:
                 ms.sort_state(best_hyp_indices)
 
+        # (9) Sort the hypotheses within each sentence (normalization for finished hyps breaks the sort)
+        for sentno in range(self.batch_size):
+            rows = slice(sentno * self.beam_size, (sentno + 1) * self.beam_size)
+            best_hyp_indices[rows] = rows.start + mx.nd.argsort(scores_accumulated[rows, 0])
+
+        # Make corresponding updates after having sorted
+        sequences = mx.nd.take(sequences, best_hyp_indices)
```

Comment:
these takes are not cheap. Having to do them twice in every search iteration is costly. Is there any way to avoid this?

Reply:
They're only done once per iteration. The final one here takes place after the loop has exited. I can change it to use a top-level operation instead of iterating over the batch (good idea).

Reply:
On the subject of takes, though — if they are indeed expensive, we might be able to get some gain by packing all of these (beam * batch, 1) tracking tensors (scores_accumulated, finished, lengths) into a single array (perhaps indexed with informative constants or as slices), in which case we'd only need a single `take()`. This would apply in the old code as well.

Reply:
That could be an idea, but let's do that in a separate commit.
Also, thanks for clarifying that this is outside of the main loop. I read the indentation wrong. In this case this should be fine I think.